### PR TITLE
add --create-manifest-if-needed flag

### PR DESCRIPTION
**Pull request by JL03-Yue**
#15254 

Add `--create-manifest-if-needed` flag
Create tool manifest automatically when installing a new local tool if the flag is on and no manifest exists.

### CODE REVIEW

File: src/Cli/Microsoft.DotNet.Cli.Utils/Constants.cs [hunk 1]
```diff
@@ -10,12 +10,14 @@ public static class Constants
         public const string DefaultConfiguration = "Debug";
 
         public static readonly string ProjectFileName = "project.json";
+        public static readonly string ManifestFileName = "dotnet-tools.json";
```
Comment (dsplaisted) [inline src/Cli/Microsoft.DotNet.Cli.Utils/Constants.cs:13]: Name suggestion:

```suggestion
        public static readonly string ToolManifestFileName = "dotnet-tools.json";
```

File: src/Cli/dotnet/ToolManifest/ToolManifestFinder.cs [hunk 2]
```diff
@@ -153,7 +155,7 @@ public bool TryFind(ToolCommandName toolCommandName, out ToolManifestPackage too
             }
         }
 
-        public FilePath FindFirst()
+        public FilePath FindFirst(bool createManifestFileOption = false)
```
Comment (dsplaisted) [inline src/Cli/dotnet/ToolManifest/ToolManifestFinder.cs:158]: Naming suggestion:

```suggestion
        public FilePath FindFirst(bool createIfNotFound = false)
```

File: src/Cli/dotnet/ToolManifest/ToolManifestFinder.cs [hunk 3]
```diff
@@ -162,12 +164,48 @@ public FilePath FindFirst()
...
-                string.Format(LocalizableStrings.ListOfSearched,
-                    string.Join(Environment.NewLine,
-                        EnumerateDefaultAllPossibleManifests().Select(f => "\t" + f.manifestfile.Value))));
+                    LocalizableStrings.CannotFindAManifestFile,
+                    string.Format(LocalizableStrings.ListOfSearched,
+                        string.Join(Environment.NewLine,
+                            EnumerateDefaultAllPossibleManifests().Select(f => "\t" + f.manifestfile.Value))));
+        }
+
+        private DirectoryPath GetInsertManifestFilePath()
```
Comment (dsplaisted) [inline src/Cli/dotnet/ToolManifest/ToolManifestFinder.cs:182]: I'm not sure what a good name for this method would be, but I think this is a bit of an improvement:

```suggestion
        private DirectoryPath GetDirectoryToCreateToolManifest()
```

**Review by dsplaisted — COMMENTED** [review]
This looks good.  The main thing still needed is to add test coverage.

File: src/Cli/dotnet/ToolManifest/LocalizableStrings.resx [hunk 4]
```diff
@@ -169,4 +169,11 @@ For a list of locations searched, specify the "-d" option before the tool name.<
...
   </data>
-</root>
+  <data name="ManifestFileContent" xml:space="preserve">
+    <value>{
+  "version": 1,
+  "isRoot": true,
+  "tools": {}
+}</value>
+  </data>
+</root>
```
[the remark this answers is not in the document]
Comment (JL03-Yue) [reply src/Cli/dotnet/ToolManifest/LocalizableStrings.resx:179]: So can it be set as a member variable in the class?

File: src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallCommandParser.cs [hunk 5]
```diff
@@ -38,6 +38,11 @@ internal static class ToolInstallCommandParser
             ArgumentHelpName = LocalizableStrings.FrameworkOptionName
         };
 
+        public static readonly Option<bool> CreateManifestIfNeededOption = new Option<bool>("--create-manifest-if-needed", LocalizableStrings.CreateManifestIfNeededOptionDescription)
+        {
+            ArgumentHelpName = LocalizableStrings.CreateManifestIfNeededOptionName
```
[the remark this answers is not in the document]
Comment (baronfel) [reply src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallCommandParser.cs:43]: No - we don't do this for the names of boolean options, only descriptions.

**Comment by nagilson** [comment]
@baronfel Was a final decision reached on this change? I believe we are going to go with what this PR does for now?

<hr>

There is the proposed ` --manifest-file`, but that was not added here to my knowledge.

<hr>

In addition, regarding:
> --create-manifest-if-needed false

Also, one question about the below proposal:
> change the default so that you don't need to specify the --create-manifest-if-needed

I assume that with the below stipulation, this means it's by default IFF the following is true:

> If the --create-manifest-if-needed flag is not specified, then we probably do not want to create a manifest in the current folder

File: src/Cli/dotnet/commands/dotnet-tool/install/LocalizableStrings.resx [hunk 6]
```diff
@@ -232,4 +232,10 @@ If you would like to create a manifest, use `dotnet new tool-manifest`, usually
   <data name="PrereleaseAndVersionAreNotSupportedAtTheSameTime" xml:space="preserve">
     <value>The --prerelease and --version options are not supported in the same command</value>
   </data>
+  <data name="CreateManifestIfNeededOptionDescription" xml:space="preserve">
+    <value>Create tool manifest when installing tools if manifest not exist.</value>
```
Comment (dsplaisted) [inline src/Cli/dotnet/commands/dotnet-tool/install/LocalizableStrings.resx:236]: @baronfel Any suggestions for this option description?  Should we describe how we decide where to put the manifest file?

File: src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallCommandParser.cs [hunk 7]
```diff
@@ -38,6 +38,11 @@ internal static class ToolInstallCommandParser
...
+        };
```
Comment (dsplaisted) [inline src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallCommandParser.cs:44]: Per @baronfel's response, we don't need to specify the ArgumentHelpName here:

```suggestion
        public static readonly Option<bool> CreateManifestIfNeededOption = new Option<bool>("--create-manifest-if-needed", LocalizableStrings.CreateManifestIfNeededOptionDescription);
```

File: src/Cli/dotnet/commands/dotnet-tool/install/LocalizableStrings.resx [hunk 8]
```diff
@@ -232,4 +232,10 @@ If you would like to create a manifest, use `dotnet new tool-manifest`, usually
...
+  </data>
+  <data name="CreateManifestIfNeededOptionName" xml:space="preserve">
+    <value>MANIFEST</value>
+  </data>
```
Comment (dsplaisted) [inline src/Cli/dotnet/commands/dotnet-tool/install/LocalizableStrings.resx:240]: Looks like we don't need this string and can delete it (see comment in ToolInstallCommandParser.cs).

```suggestion
```

**Review by dsplaisted — APPROVED** [review]
Looks great!  A few suggestions still.

File: src/Cli/dotnet/commands/dotnet-tool/install/LocalizableStrings.resx (hunk 6 above)
Comment (baronfel) [reply src/Cli/dotnet/commands/dotnet-tool/install/LocalizableStrings.resx:236]: How about "Create a tool manifest if one isn't found during tool installation. For information on how manifests are located, see [HELP URL]".

We should document the search algorithm on https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools#install-a-local-tool, though. Potentially on the [CannotFindAManifestFile](https://github.com/dotnet/sdk/pull/31231/files/6177c345e89b2357ab290c0d541fe55e209733e0#diff-4288e2d70efc0eba5c7ffec9c84313f69b4111f7611d00345c58442ef96b690aR120-R123) error as well.

Comment (dsplaisted) [reply src/Cli/dotnet/commands/dotnet-tool/install/LocalizableStrings.resx]: How about https://aka.ms/dotnet-local-tool-create-manifest-if-needed?

Comment (baronfel) [reply src/Cli/dotnet/commands/dotnet-tool/install/LocalizableStrings.resx]: An aka.ms link would be perfect - folks have been 'namespacing' them using url path segments, though - maybe something like `aka.ms/dotnet/tools/create-manifest-if-needed`?

**Comment by baronfel** [comment]
@JL03-Yue we need to doc this! I just went looking and couldn't find it. I'll make a docs issue and we can work on doing that.

**Pull request merged**